File: python/HugeFiboNumMod.py

from random import *

def naive(n, m):
	if n <= 1:
		return n

	previous = 0
	current = 1
	
	for i in range(2, n + 1):
		previous, current = current, previous + current

	return current % m

def fast(n, m):
	if n <= 1:
		return n
	
	arr = []
	arr.append(0)
	arr.append(1)
	i = 2
	while 1:
		arr.append( (arr[i - 1] + arr[i - 2]) % m )

		if arr[i - 1] == 0 and arr[i] == 1:
			break
		else:
			i += 1
	
	return arr[n % (i - 1)]
		

n, m = map(int, input().split())
# print(naive(n, m))
print(fast(n, m))

'''
N, M = map(int, input().split())

while 1:
	n = randint(1, N)
	m = randint(2, M)
	if naive(n, m) == fast(n, m):
		print(n, m, " - ok")
	else:
		print(n, m, " - fail")
		break
'''

File: java/src/HugeFiboNumMod.java
import java.util.*;

public class HugeFiboNumMod {

	private static int naive(long n, int m) {
		if (n <= 1)
			return (int) n;
		
		int previous = 0;
		int current = 1;
		int tmp = 0;
		for (long i = 2 ; i < n + 1 ; i++) {
			tmp = previous;
			previous = current;
			current = (tmp + current)  % m;
		}

		return current;
	}
	
	private static int fast(long n, int m) {
		if (n <= 1)
			return (int) n;
		
		List<Integer> arr = new ArrayList<Integer>();
		arr.add(0);
		arr.add(1);
		
		int i = 2;
		while(true) {
			arr.add( (arr.get(i - 1) + arr.get(i - 2)) % m );
			if (arr.get(i - 1) == 0 && arr.get(i) == 1)
				break;
			else
				i++;
		}
		
		int idx = (int) (n % (i - 1));		
		return arr.get( idx );
	}
	
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		
		long n = scanner.nextLong();
		int m = scanner.nextInt();
		// System.out.println(naive(n, m));
		System.out.println(fast(n, m));
		
		/*Random rand = new Random();
		while(true) {
			long i = 1 + (long) (Math.random() * n);
			int j = rand.nextInt(m) + 2;
			System.out.print(i + "/" + j);

			if (naive(i, j) == fast(i, j)) {
				System.out.println(" - ok");
			} else {
				System.out.println(" - fail");
				break;
			}
		}*/
		
		scanner.close();
	}
}
